src/rgw/rgw_redis_driver.cc

```diff
@@ -0,0 +1,775 @@
+
+      client.hdel(entry, redisFields, [&result](cpp_redis::reply &reply) {
+        if (reply.is_integer()) {
+          result = reply.as_integer();
+        }
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+
+      return result - 1;
+    } catch(std::exception &e) {
+      return -1;
+    }
+  }
+
+  ldpp_dout(dpp, 20) << "RGW Redis Cache: Object is not in cache." << dendl;
+  return -2;
+}
+
+std::string RedisDriver::get_attr(const DoutPrefixProvider* dpp, const std::string& key, const std::string& attr_name) 
+{
+  std::string entry = partition_info.location + key;
+  std::string attrValue;
+
+  if (!client.is_connected()) 
+    find_client(dpp);
+
+  if (key_exists(dpp, key)) {
+    int exists = -2;
+    std::string getValue;
+
+    /* Ensure field was set */
+    try {
+      client.hexists(entry, attr_name, [&exists](cpp_redis::reply& reply) {
+	if (!reply.is_null()) {
+	  exists = reply.as_integer();
+	}
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+    } catch(std::exception &e) {
+      return {};
+    }
+    
+    if (!exists) {
+      ldpp_dout(dpp, 20) << "RGW Redis Cache: Attribute was not set." << dendl;
+      return {};
+    }
+
+    /* Retrieve existing value from cache */
+    try {
+      client.hget(entry, attr_name, [&exists, &attrValue](cpp_redis::reply &reply) {
+	if (!reply.is_null()) {
+	  attrValue = reply.as_string();
+	}
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+    } catch(std::exception &e) {
+      return {};
+    }
+  } else {
+    ldpp_dout(dpp, 20) << "RGW Redis Cache: Object is not in cache." << dendl;
+    return {};
+  }
+
+  return attrValue;
+}
+
+int RedisDriver::set_attr(const DoutPrefixProvider* dpp, const std::string& key, const std::string& attr_name, const std::string& attrVal) 
+{
+  std::string entry = partition_info.location + key;
+  int result = 0;
+    
+  if (!client.is_connected()) 
+    find_client(dpp);
+    
+  if (key_exists(dpp, key)) {
+    /* Every attr set will be treated as new */
+    try {
+      client.hset(entry, attr_name, attrVal, [&result](cpp_redis::reply& reply) {
+	if (!reply.is_null()) {
+	  result = reply.as_integer();
+	}
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+    } catch(std::exception &e) {
+      return -1;
+    }
+  } else {
+    ldpp_dout(dpp, 20) << "RGW Redis Cache: Object is not in cache." << dendl;
+    return -2; 
+  }
+
+  return result - 1;
+}
+
+std::unique_ptr<CacheAioRequest> RedisDriver::get_cache_aio_request_ptr(const DoutPrefixProvider* dpp) 
+{
+  return std::make_unique<RedisCacheAioRequest>(this);  
+}
+
+rgw::AioResultList RedisDriver::get_async(const DoutPrefixProvider* dpp, optional_yield y, rgw::Aio* aio, const std::string& key, off_t ofs, uint64_t len, uint64_t cost, uint64_t id) 
+{
+  rgw_raw_obj r_obj;
+  r_obj.oid = key;
+
+  return aio->get(r_obj, rgw::Aio::cache_read_op(dpp, y, this, ofs, len, key), cost, id);
```

Comment:
instead of calling `rgw::Aio::cache_read_op()` for this, i think you could just wrap the `conn.async_exec()` call in a static `Aio::OpFunc` like this:
```c++
static Aio::OpFunc redis_read_op(optional_yield y, connection& conn,
                                 off_t read_ofs, off_t read_len, const std::string& key)
{
  return [y, &conn, read_ofs, read_len, key] (Aio* aio, AioResult& r) mutable {
    using namespace boost::asio;
    async_completion<yield_context, void()> init(y.get_yield_context());
    auto ex = get_associated_executor(init.completion_handler);

    request req;
    req.push("GETRANGE", key, read_ofs, read_ofs + read_len);

    auto s = std::make_unique<state>;
    auto& resp = s->resp; // save the reference because we move out of s
    conn.async_exec(req, resp, bind_executor(ex, RedisDriver::libaio_handler{aio, r, std::move(s)}));
  };
}
```